File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int prefix = 1, suffix =1;
        int ans = -1e9;
        for(int i=0;i<nums.size();i++){
            prefix *= nums[i];
            suffix *= nums[nums.size()-1-i];
            ans = max({ans, prefix, suffix});
            if(prefix == 0)prefix = 1;
            if(suffix == 0)suffix = 1;
        }
        return ans;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> m = new HashMap<>();
        int i=-1,j=0;
        int ans = 0;
        while(j < s.length()){
            int idx = m.getOrDefault(s.charAt(j),-1);
            m.put(s.charAt(j),j);
            idx = Math.max(idx, i);
            if(i != idx)i = idx;
            ans = Math.max(j-i,ans);
            j++;
        }
        return ans;
    }
}
